File: Python/21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        l3 = ListNode(0)
        pre = l3
        while l1 != None or l2 != None:
            if l1 == None:
                curr = l2
                l2 = l2.next
            elif l2 == None:
                curr = l1
                l1 = l1.next
            elif l1.val < l2.val:
                curr = l1
                l1 = l1.next
            else:
                curr = l2
                l2 = l2.next
            pre.next = curr
            pre = curr
        return l3.next

File: 2021/21. Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {        
        ListNode *head = NULL, *prev = NULL; 
        while (l1 != NULL || l2 != NULL) {
            ListNode* curr = NULL;
            if (l1 == NULL || (l2 != NULL && l2->val < l1->val)) {                
                curr = l2;
                l2 = l2->next;            
            } else {
                curr = l1;
                l1 = l1->next;
            }            
            if (head == NULL) {
                head = curr;
                prev = curr;
            } else {
                prev->next = curr;
                prev = prev->next;
            }
        }
        return head;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(), *prev = dummy;
        while (l1 != NULL && l2 != NULL) {
            if (l1->val < l2->val) {
                prev->next = l1;
                l1 = l1->next;                
            } else {
                prev->next= l2;
                l2 = l2->next;
            }
            prev = prev->next;
        }
        if (l1 != NULL) {
            prev->next = l1;
        } else {
            prev->next = l2;
        }
        ListNode* head = dummy->next;
        delete dummy;
        return head;
    }
};